### PR TITLE
[TRTLLM-15498][perf] optimize Kimi KDA prefill convolution data flow

### PR DESCRIPTION
**Pull request by jiaganc**
@coderabbitai summary

## Description

Optimize the Kimi K3 KDA prefill convolution and cache data flow on Blackwell:

- reuse GDN's packed variable-length causal convolution for Q/K/V and update
  the live convolution pool directly;
- consume packed channel-major convolution output directly, fuse the
  post-convolution transpose with Q/K L2 normalization, and pass normalized
  Q/K to both the optimized and FLA recurrence paths;
- store only the committed `W - 1` convolution window in the live cache pool
  and synchronize speculative replay caches from that layout;
- reuse fused QKVG projections without an intermediate Q/K/V concatenation;
- update cache-manager sizing, prefix/continuation handling, disaggregated
  transfer geometry, and packed-cache parity coverage for the `W - 1` layout;
- add `test_kda_decode_op.py`, which also covers the shared convolution-cache
  frontend, to the B200 test database.

KDA single-token decode ABI and direct-pool optimizations were split into draft
PR #18251. This PR is now prefill/cache-layout only.

This is an internal data-layout optimization. It does not add a user-facing API
or an opt-in environment switch.

## Test Coverage

The prefill changes were validated before the history-only PR split as part of
the combined tree. The split retained the prefill source and test hunks
unchanged and reran commit-time validation on the prefill-only history.

- Clean SM100 (`100-real`) native build on B200: passed (3,883 Ninja steps).
- Complete `tests/unittest/_torch/modules/kimi_kda/` suite on B200: 76 passed,
  16 intentional skips, 0 failed.
- FP8 packed-prefill coverage: 3 passed, 0 failed.
- KDA disaggregated transfer coverage: 18 passed, 0 failed.
- K3 recurrent-state bounce/helix coverage: 13 passed, 0 failed.
- The B200 test database includes `test_kda_decode_op.py`; repository AST
  test-list validation completed without stale selectors.
- Earlier combined-tree B300 validation: 278 passed, 3 skipped, 3 xfailed,
  0 failed; targeted FLA fallback and fused-QKVG prefill parity reruns passed.
- Commit-time formatting, lint, DCO, conflict, and repository validation hooks
  passed on the rewritten prefill-only branch; `git diff --check` passed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: cpp/tensorrt_llm/kernels/kdaDecode/kdaDecode.h [hunk 1]
```diff
@@ -42,6 +42,21 @@ constexpr bool shouldUseCompactHeads(int batchSize, int numHeads, int numValueHe
...
+//! kernel consume column views of fused projection outputs without first
+//! materializing each view as a contiguous tensor.
+struct KdaDecodeIoLayout
+{
+    int xQRowStride;
+    int xKRowStride;
+    int xVRowStride;
+    int gateRowStride;
+    int betaRowStride;
+    int outputNormGateRowStride;
```
[the remark this answers is not in the document]
Comment (jiaganc) [reply cpp/tensorrt_llm/kernels/kdaDecode/kdaDecode.h]: Fixed in 07bc9c6e26. KdaDecodeIoLayout now stores all six row strides as int64_t, and the host-side layout construction preserves the tensor strides without narrowing casts.

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_kernels.py [hunk 2]
```diff
@@ -66,6 +68,198 @@ def is_kda_optimized_supported() -> bool:
...
+    ).to(tl.float32)
+    k /= tl.sqrt(tl.sum(k * k, axis=1) + l2_norm_eps)[:, None]
+    tl.store(k_out_ptr + output_offsets, k, mask=mask)
+
+    v = tl.load(
+        packed_ptr + 2 * projection_size * num_tokens + source_offsets,
+        mask=mask,
+        other=0.0,
+    )
+    tl.store(v_out_ptr + output_offsets, v, mask=mask)
```
[the remark this answers is not in the document]
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/_kda_kernels.py:121]: Fixed in 07bc9c6e26. The Triton kernel now converts num_tokens and both source-index components to int64, computes an int64 section_stride, and uses it for the K/V section bases.

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py [hunk 3]
```diff
@@ -424,48 +422,51 @@ def forward_prefill(
...
-            v_proj_states, cache=conv_v_in, output_final_state=True, cu_seqlens=cu_seqlens
+        # Reuse GDN's packed variable-length causal convolution. It reads
+        # and writes the live [slots, 3D, W - 1] pool directly and honors
+        # has_initial_state for fresh versus continuation requests.
+        packed_conv = qkv.squeeze(0).transpose(0, 1).contiguous()
+        assert self._packed_conv_weight is not None
+        causal_conv1d_fn(
+            packed_conv,
+            self._packed_conv_weight,
+            query_start_loc=cu_seqlens.to(torch.int32),
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py]: `Mamba2Metadata` already maintains an int32 `query_start_loc` alongside the int64 mirror, and its comment states it exists specifically because `causal_conv1d_fn` only accepts int32. Passing `query_start_loc_long` and casting it back here allocates and copies once per layer per step. Use `mamba_metadata.query_start_loc[:num_prefills + 1]` directly - it is non-None whenever `num_contexts > 0`, which is exactly when this path runs.

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py [hunk 4]
```diff
@@ -217,5 +223,6 @@ def run_kda_decode_fusion_cuda(
         lower_bound_value,
         float(scale),
         float(onorm_eps),
+        get_env_enable_pdl(),
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py]: `get_env_enable_pdl()` re-reads `os.environ` on every call, while the C++ `getEnvEnablePDL()` caches the same value with `call_once`. Consider resolving it once in `finalize_decode_weights` and storing it on the module, so the eager (non-graph-captured) decode path does not repeat the lookup per layer.

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py [hunk 5]
```diff
@@ -159,19 +168,16 @@ def run_kda_decode_fusion_cuda(
...
         if not (
-            cs_q.stride(1) == 1
-            and cs_k.stride(1) == 1
-            and cs_v.stride(1) == 1
-            and cs_q.stride(2) == q_stride
-            and cs_k.stride(2) == q_stride
-            and cs_v.stride(2) == v_stride
+            cs_q.stride(0) == cs_k.stride(0) == cs_v.stride(0)
+            and cs_q.stride(0) >= min_slot_stride
+            and cs_q.stride(1) == cs_k.stride(1) == cs_v.stride(1) == 3
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py]: The literal `3` here is `W - 1`; `kdaDecodeOp.cpp` spells the same constant as `kKernelWidth - 1`. Deriving it from `cs_q.shape[2]` would keep this check tied to the actual conv width and let the clearer C++ "only convolution width 4 is supported" message be the one users see.

File: cpp/tensorrt_llm/common/envUtils.h [hunk 6]
```diff
@@ -66,8 +66,8 @@ bool getEnvEnableCascadeMmha();
 bool getEnvEnableTrtllmgenMoeRoutingRenormPDL();
 
 template <typename KernelFn, typename... Args>
-inline void launchWithPdlWhenEnabled(char const* name, KernelFn kernelFn, dim3 grid, dim3 block, size_t dynamicShmSize,
-    cudaStream_t stream, Args&&... args)
+inline void launchWithPdl(char const* name, bool enablePdl, KernelFn kernelFn, dim3 grid, dim3 block,
+    size_t dynamicShmSize, cudaStream_t stream, Args&&... args)
 {
     TLLM_LOG_DEBUG("Enable PDL in %s", name);
```
Comment (zhaoyangwang-nvidia) [inline cpp/tensorrt_llm/common/envUtils.h:72]: Now that `enablePdl` is an explicit parameter, this unconditional `TLLM_LOG_DEBUG("Enable PDL in %s")` also fires when PDL is off. Log the actual value, or guard the message with `if (enablePdl)`.

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py (hunk 5 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py]: Fixed in 603923c06b. The Python layout check now derives conv_state_width from cs_q.shape[2] and uses it for slot and section stride validation, leaving the fixed supported-width validation to C++.

File: cpp/tensorrt_llm/common/envUtils.h (hunk 6 above)
Comment (jiaganc) [reply cpp/tensorrt_llm/common/envUtils.h:72]: Fixed in 603923c06b. launchWithPdl now logs the actual enabled or disabled value instead of unconditionally saying PDL is enabled.

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py (hunk 3 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py]: Fixed in 603923c06b. forward_prefill now passes the existing int32 mamba_metadata.query_start_loc slice directly to causal_conv1d_fn; the int64 mirror remains available for KDA prefill.

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py (hunk 4 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py]: Fixed in 603923c06b. I kept launch policy out of finalize_decode_weights: KDAKernelDispatch resolves the common get_env_enable_pdl helper once during module construction and reuses the stored bool for eager decode launches. The module-level parity test injects enable_pdl directly for the off/on cases, so it no longer relies on environment mutation or cache behavior.

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py [hunk 7]
```diff
@@ -159,19 +167,18 @@ def run_kda_decode_fusion_cuda(
...
+            cs_q.stride(0) == cs_k.stride(0) == cs_v.stride(0)
+            and cs_q.stride(0) >= min_slot_stride
+            and cs_q.stride(1) == cs_k.stride(1) == cs_v.stride(1) == conv_state_width
+            and cs_q.stride(2) == cs_k.stride(2) == cs_v.stride(2) == 1
         ):
             raise ValueError(
-                "update_conv_cache expects transposed conv-state layout: "
-                "shape [B, dim, 3], stride(1)=1, stride(2)=dim"
+                "update_conv_cache expects section views of packed "
+                "[slots, 3 * dim, width] conv states"
```
[the remark this answers is not in the document]
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/_kda_decode.py]: Fixed in 525ef85f5a. The update_conv_cache path now requires cs_q, cs_k, and cs_v to be rank 3 before accessing shape[2] or stride(2), so malformed lower-rank inputs raise the intended ValueError. Validated on B200: an explicit rank-2 cs_q case raises ValueError with the rank-3 message, and test_kda_decode_op.py reports 12 passed, 14 expected SM103-only skips, and 0 failures.

**Comment by jiaganc** [comment]
I put the KDA decode change to a new PR #18251 to reduce current PR's scope.

File: tests/unittest/_torch/modeling/test_kimi_kda_fp8_packed_prefill.py [hunk 8]
```diff
@@ -110,10 +111,12 @@ def test_fp8_packed_qkv_prefill_matches_separate_path_and_updates_state(
     metadata = SimpleNamespace(
         use_initial_states=use_initial_states,
         has_initial_states=torch.tensor(has_initial_states, device="cuda", dtype=torch.bool),
+        state_indices=slot_indices.to(torch.int32),
+        query_start_loc=cu_seqlens.to(torch.int32),
     )
     hidden = torch.randn(num_tokens, _Cfg.hidden_size, device="cuda", dtype=torch.bfloat16) * 0.05
     hidden_pristine = hidden.clone()
-    conv_seed = torch.randn(slots, 3 * d, conv_size, device="cuda", dtype=torch.bfloat16) * 0.02
+    conv_seed = torch.randn(slots, 3 * d, conv_size - 1, device="cuda", dtype=torch.bfloat16) * 0.02
```
[the remark this answers is not in the document]
Comment (jiaganc) [reply tests/unittest/_torch/modeling/test_kimi_kda_fp8_packed_prefill.py:119]: Task [TRTLLM-14818](https://jirasw.nvidia.com/browse/TRTLLM-14818) will add tests to QA list.

**Review by nvpohanh — APPROVED** [review]

**Review by reasonsolo — APPROVED** [review]
tests/unittest/disaggregated looks good to me

File: tests/unittest/_torch/modeling/test_kimi_kda_fp8_packed_prefill.py (hunk 8 above)
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_kimi_kda_fp8_packed_prefill.py:119]: Out of curiosity, why the `- 1` now?

**Review by 2ez4bz — APPROVED** [review]
Approving changes to `tests/unittest/_torch/modeling/test_kimi_kda_fp8_packed_prefill.py`.

File: tests/unittest/_torch/modeling/test_kimi_kda_fp8_packed_prefill.py (hunk 8 above)
Comment (jiaganc) [reply tests/unittest/_torch/modeling/test_kimi_kda_fp8_packed_prefill.py:119]: `conv_seed` is used to generated the KDA conv cache for testing, and in this PR we change the conv cache shape from `conv_size` to `conv_size - 1`. KDA only needs 4 history tokens' qkv and 1 latest token's qkv to complete the 5 width conv.

**Review by Shixiaowei02 — APPROVED** [review]
